### ISSUE TITLE
Third `beforeEach` not being called before third `it`

### ISSUE THREAD

**Issue by Genhain** [Quick/Quick#775]
- [X] I have read [CONTRIBUTING](https://github.com/Quick/Quick/blob/master/CONTRIBUTING.md) and have done my best to follow them.

### What did you do?

Created a protocol testing framework to use in which to write my test cases, making it easier to switch to and from various libraries if i wish. It was done in the style of Cucumber. 

```swift
protocol GivenWhenThenable {
    func feature(_ description: String, closure: () -> Void)
    func scenario(_ description: String, closure: () -> Void)
    func given(_ description: String, file: String, line: UInt, closure: @escaping () -> Void)
    func and(_ description: String, file: String, line: UInt, closure: @escaping () -> Void)
    func but(_ description: String, file: String, line: UInt, closure: @escaping () -> Void)
    func when(_ description: String, file: String, line: UInt, closure: @escaping () -> Void)
    func then(_ description: String, file: String, line: UInt, closure: @escaping () -> Void)
}
```
QuickSpec extension
```swift
extension QuickSpec: GivenWhenThenable {
    
    typealias BDDClosure = (description: String, file: String, line: UInt, closure: () -> Void)
    
    private enum AssociatedKeys {
        static var setupClosures = "QuickSpec.setupClosures"
    }
    
    var setupClosures: [BDDClosure]? {
        get {
            guard let setupClosures = objc_getAssociatedObject(self, &AssociatedKeys.setupClosures) as? [BDDClosure] else {
                return []
            }
            return setupClosures
        }
        set {
            if let newValue = newValue {
                objc_setAssociatedObject(self,
                                         &AssociatedKeys.setupClosures,
                                         newValue as [BDDClosure]?,
                                         .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
            }
        }
    }
    
    func feature(_ description: String, closure: () -> Void) {
        describe(description, closure: closure)
    }
    
    func scenario(_ description: String, closure: () -> Void) {
        context(description, closure: closure)
    }
    
    func given(_ description: String, file: String = #file, line: UInt = #line, closure: @escaping () -> Void) {
        setupClosures?.append((description, file, line, closure))
    }
    
    func and(_ description: String, file: String = #file, line: UInt = #line, closure: @escaping () -> Void) {
        setupClosures?.append((description, file, line, closure))
    }
    
    func but(_ description: String, file: String = #file, line: UInt = #line, closure: @escaping () -> Void) {
        setupClosures?.append((description, file, line, closure))
    }
    
    func when(_ description: String, file: String = #file, line: UInt = #line, closure: @escaping () -> Void) {
        setupClosures?.append((description, file, line, closure))
    }
    
    func then(_ description: String, file: String = #file, line: UInt = #line, closure: @escaping () -> Void) {
        
        var testDescription: String = ""
        
        if let descriptions = (setupClosures?.map { $0.description }) {
            testDescription = descriptions.joined(separator: " ")
        }
        
        testDescription += " " + description
        
        it(testDescription, flags: [:], file: file, line: line) {
            self.setupClosures?.forEach { $0.closure() }
            closure()
            self.setupClosures = nil
        }
    }
}
```
it generally seems to works with the exception of trying to use `beforeEach` with the way that i have formatted them

```swift
class PartyTests: QuickSpec {
 
    override func spec() {

        beforeEach {
            MagicalRecord.setupCoreDataStackWithInMemoryStore()
        }
        
        afterEach {
            MagicalRecord.cleanUp()
        }
        
        feature("session Predicate") {
            session_predicate_tests()
        }

        feature("get name Badge Info with index") {
            name_badge_information_for_index_tests()
        }
    }
}
```

The tests inside the functions
```swift
func session_predicate_tests() {
    var sessionPredicate: NSPredicate?
    var SUT: Party!
    
    beforeEach {
        SUT = Party.mr_createEntity()
    }
    
    afterEach {
        SUT.mr_deleteEntity()
        SUT = nil
    }
    
    scenario("is a session") {
        
        let expected = NSNumber.random(using: &Xoroshiro.default)
        given("session ID is set") {
            SUT.sessionId = expected
        }
        
        when("you query the party session predicate") {
            sessionPredicate = SUT.partySessionPredicate()
        }
        
        then("it should have format \"sessionId == \(expected)\"") {
            expect(sessionPredicate?.predicateFormat).to(equal("sessionId == \(expected)"))
        }
    }
}

func name_badge_information_for_index_tests() {
    var SUT: Party!
    
    beforeEach {
        SUT = Party.mr_createEntity()
    }
    
    afterEach {
        SUT.mr_deleteEntity()
        SUT = nil
    }
    
    var actual: [AnyHashable: Any]!
    var expected: Int!
    let nameBadgeStub =
    """
        {
            "name_badge":
            {
                "stub_data":"stub"
            }
        }
    """

    scenario("there is 3 badges") {
        
        given("the badges are ascending by ID with the middle badge having a random ID") {
            expected = 111
            SUT.nameBadge = nameBadgeStub
           
            SUT.nameBadges = [["id": expected, "design":[:]],
                              ["id": Int.random(in: expected..<Int.max, using: &Xoroshiro.default)!, "design":[:]],
                              ["id": Int.max, "design":[:]]]
        }
        
        when("user attempts to print calling this function with an index of 0 which does not exist") {
            actual = SUT.nameBadgeInfo(forIndex: 0)
        }
        
        then("function should return lowest id which is \(expected)") {
            expect(actual["id"] as? Int).to(equal(expected))
        }
    }
}
```

### What did you expect to happen?

when using `beforeEach` it should call it before every `given`, `when`, `then` i have specified preferably isolated to each describe or context group i have assigned. in this case

the first `beforeEach ` which calls ever `MagicalRecord.setupCoreDataStackWithInMemoryStore()`
followed by the `session_predicate_tests()` `beforeEach` when running the `then` followed by the `name_badge_information_for_index()` `beforeEach` when running the `then` 

### What actually happened instead?

The first two aforementioned `beforeEach` blocks get called but the last one does not before attempting to assign to the attributes of the SUT in the `given` block of the scenario "there is 3 badges" which throws an exception as the SUT has not been initialised yet.

### Environment

List the software versions you're using:

 - Quick: *1.2.0*
 - Nimble: *7.0.2*
 - Xcode Version: *9.2 (9C40b)* 
 - Swift Version: *Xcode Default*

Please also mention which package manager you used and its version. Delete the
other package managers in this list:

 - Cocoapods: *1.5.0*

### Project that demonstrates the issue
[Prototype Example project](https://github.com/Genhain/QuickExample)

**Comment by jwfriese** [comment]
@Genhain - A few questions that will help us diagnose the problem in the absence of an example project:

1) Does the assignment fail in the same way for `scenario("is a session")` -> `given("session ID is set")`?
2) Does the crash happen on `SUT.nameBadge = nameBadgeStub`?
3) Is it possible that the assignment in the `beforeEach` fails? (i.e. did you verify with a breakpoint or a `print(...)` that the `beforeEach` is indeed not running?) 
4) Does this happen when you run `feature("get name Badge Info with index")` by itself? 
5) Does this happen when you delete the `afterEach` within the offending `feature`?

**Comment by Genhain** [comment]
@jwfriese Any tips on where i can upload a working example project for Xcode for you guys to use?

1. No that example runs fine it's before each is called no problem
2. Correct, because the beforeEach has not been called for that Given, When, Then
3. It does not fail, it's just is not called in the correct order, if i remove any code that could throw an exception, eventually that `beforeEach` does get called
4. No, it runs in the correct order on its own.
5. Yes, `afterEach` does not affect the erroneous behaviour im afraid.

**Comment by Genhain** [comment]
@jwfriese [Heres](https://github.com/Genhain/QuickExample) a repo for the example.

**Comment by jwfriese** [comment]
@Genhain - I think I've got it.

**TL;DR - Quick collects _all_ the DSL elements (all the function calls within `describe` and `context`) before even executing the first `it`. Your implementation of `setupClosures` on the `QuickSpec` extension does not break up the `given`s, `when`s, and `then`s by their execution group, and so _all_ of them are set to execute for _every_ `it` that lives within a `QuickSpec` implementation.** Alas, not every `beforeEach` executes for each `it` -- those are scoped by execution group. 

Longer explanation follows, if you're interested in more detail.

I'm going to do my best to illustrate the way that Quick executes its control flow so that the problem with your extension becomes apparent. 

Consider the following example, which I tried to plan so that it would as closely replicate the original problem example as much as possible. 

```
class ExampleSpec: QuickSpec {
  override func spec() {
        beforeEach {
            print("spec->beforeEach")
        }

        afterEach {
            print("spec->afterEach")
        }

        describe("session Predicate") {
            beforeEach {
                print("spec->describe(session Predicate)->beforeEach")
            }

            context("is a session") {
                beforeEach {
                    print("spec->describe(session Predicate)->context(is a session)->beforeEach")
                }

                it("it should have format sessionId == expected") {
                    print("spec->describe(session Predicate)->context(is a session)->it")
                }
            }
        }

        describe("get name Badge Info with index") {
            beforeEach {
                print("spec->describe(get name Badge Info with index)->beforeEach")
            }

            context("there is 3 badges") {
                beforeEach {
                    print("spec->describe(get name Badge Info with index)->context(there is 3 badges)->beforeEach")
                }

                it("function should return lowest id which is expected") {
                    print("spec->feature(get name Badge Info with index)->scenario(there is 3 badges)->it")
                }
            }
        }
    }
}
```

This _roughly_ mirrors what's going on with your example using your Cucumber-like extensions and framework. Now, when Quick runs through this, it starts at the top at the invocation of `spec` and runs through the entire function call, loading up the `beforeEach`s, `afterEach`s, `describe`s, `context`s, and `it`s into something like a tree.

```
                                                             spec
                                                                |
                                                       beforeEach
                                              /                                   \
                       des("session...")                            des("get name Badg...")
                                   |                                                            |
                         beforeEach                                           beforeEach
                                   |                                                            |
                  context("is a sessi...")                         context("there is 3 b...")
                                  |                                                             |
                         beforeEach                                           beforeEach
                                  |                                                             |
              it("it should have format...")                 it("function should retur...")
```

Now, our example does not have many branches, because there are not many nested describes or contexts and each one only has a single `it`, but, hopefully, you get the picture. 

Each of these full journeys down the tree, from `spec` to `it`, is a full execution. Now, in the next picture, I am going to number the `beforeEach`s, and then refer to the numbering when making the next point:
```
                                                             spec
                                                                |
                                                       beforeEach
                                              /                                   \
                       des("session...")                            des("get name Badg...")
                                   |                                                            |
                         beforeEach_1                                           beforeEach_2
                                   |                                                            |
                  context("is a sessi...")                         context("there is 3 b...")
                                  |                                                             |
                         beforeEach_3                                          beforeEach_4
                                  |                                                             |
              it("it should have format...")                 it("function should retur...")
```

As you're probably familiar, the execution of `it("it should have format...")` will _only_ execute `beforeEach_1` and `beforeEach_3`, in that order. 

Likewise, `it("function should retur...")` will _only_ execute `beforeEach_2` and `beforeEach_4`, in that order. In other words, these `beforeEach` executions are naturally scoped to the execution of the `it` that occupies their side of the test subtree that Quick builds. 

Now, on to your problem. Take a look at the how your extension methods make use of what is effectively a global list in the form of the extension property `setupClosures`:
```
func given(_ description: String, file: String = #file, line: UInt = #line, closure: @escaping () -> Void) {
        setupClosures?.append((description, file, line, closure))
    }
    
    func and(_ description: String, file: String = #file, line: UInt = #line, closure: @escaping () -> Void) {
        setupClosures?.append((description, file, line, closure))
    }
    
    func but(_ description: String, file: String = #file, line: UInt = #line, closure: @escaping () -> Void) {
        setupClosures?.append((description, file, line, closure))
    }
    
    func when(_ description: String, file: String = #file, line: UInt = #line, closure: @escaping () -> Void) {
        setupClosures?.append((description, file, line, closure))
    }
```

Each of them appends their execution onto a list, _as soon as Quick hits it_. That is, before we ever execute a single `it` block. Now, let's look at how all those elements in `setupClosures` finally get used:
```
func then(_ description: String, file: String = #file, line: UInt = #line, closure: @escaping () -> Void) {
        
        var testDescription: String = ""
        
        if let descriptions = (setupClosures?.map { $0.description }) {
            testDescription = descriptions.joined(separator: " ")
        }
        
        testDescription += " " + description
        
        it(testDescription, flags: [:], file: file, line: line) {
            self.setupClosures?.forEach { $0.closure() }
            closure()
            self.setupClosures = nil
        }
    }
```

Can you see what might be causing the issue? Whenever we execute a `then` in this Cucumber-like framework, we execute *all* of those `setupClosures` that we have collected up to that point, _regardless of which test execution subtree we're running in_. 

To go back to our illustration, what I think you're trying to create is an execution tree that looks like this:
```
                                                             spec
                                                                |
                                                       beforeEach
                                              /                                   \
                    feature("session...")                        feature("get name Badg...")
                                   |                                                            |
                         beforeEach_1                                        beforeEach_2
                                   |                                                            |
                  scenario("is a sessi...")                      scenario("there is 3 b...")
                                  |                                                             |
                             given_1                                                    given_2
                                  |                                                             |
                             when_1                                                    when_2
                                  |                                                             |
           then("it should have format...")             then("function should retur...")
```

but, in reality, because *all* the `given`s and `when`s for the entire tree are executed every time a `then` executes within a subtree, what we end up with is a testing tree that looks more like this:
```
                                                             spec
                                                                |
                                                       beforeEach
                                              /                                   \
                    feature("session...")                        feature("get name Badg...")
                                   |                                                            |
                         beforeEach_1                                         beforeEach_2
                                   |                                                            |
                  scenario("is a sessi...")                      scenario("there is 3 b...")
                                  |                                                             |
                             given_1                                                  given_1
                                  |                                                             |
                             when_1                                                  when_1
                                  |                                                             |
                             given_2***                                            given_2
                                  |                                                             |
                             when_2                                                 when_2
                                  |                                                             |
           then("it should have format...")             then("function should retur...")
```
I've starred (***) above where we're exploding. `given_2` is intended to execute after `beforeEach_2` does some setup for its closure, but its executing within the testing path that follows `beforeEach_1` instead. 

I hope that this helps and does not confuse things further @Genhain. Let me know if you have any questions or need any further clarification.

**Comment by jwfriese** [comment]
@Genhain - I agree with that conclusion. I think that Quick might not have a suitable interface to make it easy to achieve a Cucumber-like experience. 

If you tinker at all with Quick and find you're able to make some progress, please feel free to continue the conversation here or in another issue!

**Comment by Genhain** [comment]
I would have liked to leverage some of the functionality that quick implements already. Alas I may have to try and emulate it myself, if I manage to get it working with quick or not I will I will report back my findings.
